uber/automated_emails.py

```diff
@@ -23,53 +25,210 @@ def __init__(self, model, subject, template, filter, *, sender=None, extra_data=
+        model = getattr(model_instance, 'email_model_name', model_instance.__class__.__name__.lower())
+        return render('emails/' + self.template, dict({model: model_instance}, **self.extra_data))
 
-    def render(self, x):
-        model = getattr(x, 'email_model_name', x.__class__.__name__.lower())
-        return render('emails/' + self.template, dict({model: x}, **self.extra_data))
+    def send(self, model_instance, raise_errors=True):
+        """
+        Actually send an email to a particular model instance (i.e. a particular attendee).
 
-    def send(self, x, raise_errors=True):
+        Doesn't perform any kind of checks at all if we should be sending this, just immediately sends the email
+        no matter what.
+
+        NOTE: use attempt_to_send() instead of calling this directly if you don't 100% know what you're doing.
+        """
         try:
             format = 'text' if self.template.endswith('.txt') else 'html'
-            send_email(self.sender, x.email, self.subject, self.render(x), format, model=x, cc=self.cc, ident=self.ident)
+            send_email(self.sender, model_instance.email, self.subject,
+                       self.render(model_instance), format,
+                       model=model_instance, cc=self.cc, ident=self.ident)
         except:
             log.error('error sending {!r} email to {}', self.subject, x.email, exc_info=True)
             if raise_errors:
                 raise
 
+    @property
+    def when_txt(self):
+        """
+        Return a textual description of when the date filters are active for this email category.
+        """
+
+        return '\n'.join([filter.active_when for filter in listify(self.when)])
+
+    @classmethod
+    def get_approved_subjects(cls, session):
+        return {ae.subject for ae in session.query(ApprovedEmail)}
+
+    @classmethod
+    def get_previously_sent_emails(cls, session):
+        return set(session.query(Email.model, Email.fk_id, Email.ident))
+
     @classmethod
     def send_all(cls, raise_errors=False):
-        if not c.AT_THE_CON and (c.DEV_BOX or c.SEND_EMAILS):
-            with Session() as session:
-                approved = {ae.subject for ae in session.query(ApprovedEmail)}
-                all_sent = set(session.query(Email.model, Email.fk_id, Email.ident))
-                for model, lister in cls.queries.items():
-                    for inst in lister(session):
-                        sleep(0.01)  # throttle CPU usage
-                        for rem in cls.instances.values():
-                            if isinstance(inst, rem.model) and (not rem.needs_approval or rem.subject in approved):
-                                if rem.should_send(inst, all_sent):
-                                    rem.send(inst, raise_errors=raise_errors)
+        """
+        Do a run of our automated email service.  This function is called once every couple of minutes.
+        """
+        SendAllAutomatedEmailsJob().run(raise_errors)
```

Comment:
OK, this ended up being a much bigger refactor than should really be in this pull request BUT I was really having serious trouble parsing this giant send_all() loop, and considered it kind of a bad sign because I've been swimming in the guts of this email sending system for a long time now.

I took this ```send_all()``` call and refactored it out into a new class named ```SendAllAutomatedEmailsJob```.  I also went berserk overkill with the documentation because for some reason I always forget what all is going on in the middle of reading this.  I think this really nicely separates out the AutomatedEmail stuff from the daemon stuff, and the end result is a lot cleaner and easier to read, despite being a bit more verbose.

While doing this it occured to me that (in a *different* PR) we should probably rename AutomatedEmail to AutomatedEmailCategory.  AutomatedEmail to me always sounded like it could be describing i.e. a specific email sent to a specific attendee, whereas these are really describing email categories i.e. "the registration confirmation email category"

Anyway, I have also incorporated the other changes from the code review and this PR is pretty much ready to go.  I definitely want to test it a bunch and write some new unit tests for the ```when``` classes, but beyond that, it's pretty ready.